integration/kube_integration_test.go

```diff
@@ -1611,3 +1612,88 @@ func testKubeJoin(t *testing.T, suite *KubeSuite) {
 	require.Contains(t, participantOutput, []byte("echo hi"))
 	require.Contains(t, out.String(), []byte("echo hi2"))
 }
+
+// testKubeConnectionLimit checks that the `max_kubernetes_connections` role option is enforced.
+func testKubeConnectionLimit(t *testing.T, suite *KubeSuite) {
+	teleport := NewInstance(InstanceConfig{
+		ClusterName: Site,
+		HostID:      HostID,
+		NodeName:    Host,
+		Priv:        suite.priv,
+		Pub:         suite.pub,
+		log:         suite.log,
+	})
+
+	const maxConnections = 10
+	hostUsername := suite.me.Username
+	kubeGroups := []string{testImpersonationGroup}
+	kubeUsers := []string{"alice@example.com"}
+	role, err := types.NewRoleV3("kubemaster", types.RoleSpecV5{
+		Allow: types.RoleConditions{
+			Logins:     []string{hostUsername},
+			KubeGroups: kubeGroups,
+			KubeUsers:  kubeUsers,
+		},
+		Options: types.RoleOptions{
+			MaxKubernetesConnections: maxConnections,
+		},
+	})
+	require.NoError(t, err)
+	teleport.AddUserWithRole(hostUsername, role)
+
+	err = teleport.Start()
+	require.NoError(t, err)
+	defer teleport.StopAll()
+
+	// set up kube configuration using proxy
+	proxyClient, proxyClientConfig, err := kubeProxyClient(kubeProxyConfig{
+		t:          teleport,
+		username:   hostUsername,
+		kubeUsers:  kubeUsers,
+		kubeGroups: kubeGroups,
+	})
+	require.NoError(t, err)
+
+	ctx := context.Background()
+	// try get request to fetch available pods
+	pod, err := proxyClient.CoreV1().Pods(testNamespace).Get(ctx, testPod, metav1.GetOptions{})
+	require.NoError(t, err)
+
+	openExec := func() error {
+		// interactive command, allocate pty
+		term := NewTerminal(250)
+		out := &bytes.Buffer{}
+
+		return kubeExec(proxyClientConfig, kubeExecArgs{
+			podName:      pod.Name,
+			podNamespace: pod.Namespace,
+			container:    pod.Spec.Containers[0].Name,
+			command:      []string{"/bin/sh"},
+			stdout:       out,
+			tty:          true,
+			stdin:        term,
+		})
+	}
+
+	// Create and maintain the maximum amount of open connections
+	errors := make(chan error)
+	for i := 0; i < maxConnections; i++ {
+		go func() {
```

Comment:
Maybe we should have a waitgroup here to wait for all the intended sessions to be opened?

Reply:
Added a check so that we wait until all sessions are open.